File: app/src/main/java/com/wolkowiczmateusz/android/cleanmvpretrofitrxdaggerboilerplate/data/repository/datasource/OfflineDataStore.java
package com.wolkowiczmateusz.android.cleanmvpretrofitrxdaggerboilerplate.data.repository.datasource;

import com.wolkowiczmateusz.android.cleanmvpretrofitrxdaggerboilerplate.data.model.UserEntity;
import com.wolkowiczmateusz.android.cleanmvpretrofitrxdaggerboilerplate.device.sharedpreferences.UserStorage;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.Single;


public class OfflineDataStore {

    private UserStorage userStorage;

    @Inject
    public OfflineDataStore(UserStorage userStorage) {
        this.userStorage = userStorage;
    }

    public UserEntity getLoginUser() {
        return userStorage.getUser();
    }

    public Completable logout() {
        return userStorage.logout();
    }

    public Single<Boolean> isUserLogin() {
        return Single.defer(userStorage::isUserLogin);
    }

    public void saveUserToStorage(UserEntity userEntity) {
        userStorage.saveUser(userEntity);
    }
}

File: app/src/main/java/com/wolkowiczmateusz/android/cleanmvpretrofitrxdaggerboilerplate/data/repository/datasource/OfflineDataStore.kt
package com.wolkowiczmateusz.android.cleanmvpretrofitrxdaggerboilerplate.data.repository.datasource

import com.wolkowiczmateusz.android.cleanmvpretrofitrxdaggerboilerplate.device.sharedpreferences.UserStorage
import com.wolkowiczmateusz.android.cleanmvpretrofitrxdaggerboilerplate.domain.model.User
import io.reactivex.Completable
import io.reactivex.Single
import javax.inject.Inject

class OfflineDataStore @Inject
constructor(private val userStorage: UserStorage) {

    val loginUser: User
        get() = userStorage.user

    val isUserLogin: Single<Boolean>
        get() = Single.defer { userStorage.isUserLogin }

    fun logout(): Completable {
        return userStorage.logout()
    }

    fun saveUserToStorage(user: User) {
        userStorage.saveUser(user)
    }
}
